phpunit/class-wp-theme-json-resolver-test.php

```diff
@@ -358,4 +358,164 @@ public function test_get_user_data_from_wp_global_styles_does_not_use_uncached_q
+
+		// Make sure there is data from the theme origin.
+		switch_theme( 'block-theme' );
+
+		// Make sure there is data from the user origin.
+		self::register_user_data();
+
+		$theme_json = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( 'default' );
+		$settings   = $theme_json->get_settings();
+		$styles     = $theme_json->get_styles_block_nodes();
+		$this->assertTrue( isset( $settings['color']['palette']['default'] ), 'core palette is present' );
+		$block_styles = array_filter(
+			$styles,
+			function( $element ) {
+				return isset( $element['name'] ) && 'my/block-with-styles' === $element['name'];
+			}
+		);
+		$this->assertTrue( count( $block_styles ) === 0, 'block styles are not present' );
+		$this->assertFalse( isset( $settings['color']['palette']['theme'] ), 'theme palette is not present' );
+		$this->assertFalse( isset( $settings['color']['palette']['custom'] ), 'user palette is not present' );
+
+		unregister_block_type( 'my/block-with-styles' );
+	}
+
+	/**
+	 * @covers WP_Theme_JSON_Resolver::get_merged_data
+	 */
+	public function test_get_merged_data_returns_origin_blocks() {
+		// Make sure there is data from the blocks origin.
+		self::register_block_data( 'my/block-with-styles' );
+
+		// Make sure there is data from the theme origin.
+		switch_theme( 'block-theme' );
+
+		// Make sure there is data from the user origin.
+		self::register_user_data();
+
+		$theme_json = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( 'blocks' );
+		$settings   = $theme_json->get_settings();
+		$styles     = $theme_json->get_styles_block_nodes();
+		$this->assertTrue( isset( $settings['color']['palette']['default'] ), 'core palette is present' );
+		$block_styles = array_filter(
+			$styles,
+			function( $element ) {
+				return isset( $element['name'] ) && 'my/block-with-styles' === $element['name'];
+			}
+		);
+		$this->assertTrue( count( $block_styles ) === 1, 'block styles are present' );
+		$this->assertFalse( isset( $settings['color']['palette']['theme'] ), 'theme palette is not present' );
+		$this->assertFalse( isset( $settings['color']['palette']['custom'] ), 'user palette is not present' );
+
+		unregister_block_type( 'my/block-with-styles' );
+	}
+
+	/**
+	 * @covers WP_Theme_JSON_Resolver::get_merged_data
+	 */
+	public function test_get_merged_data_returns_origin_theme() {
+		// Make sure there is data from the blocks origin.
+		self::register_block_data( 'my/block-with-styles' );
+
+		// Make sure there is data from the theme origin.
+		switch_theme( 'block-theme' );
+
+		// Make sure there is data from the user origin.
+		self::register_user_data();
+
+		$theme_json = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( 'theme' );
+		$settings   = $theme_json->get_settings();
+		$styles     = $theme_json->get_styles_block_nodes();
+		$this->assertTrue( isset( $settings['color']['palette']['default'] ), 'core palette is present' );
+		$block_styles = array_filter(
+			$styles,
+			function( $element ) {
```

Comment:
Please see https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#one-off-functions-for-hooks
```suggestion
			static function( $element ) {
```